akka-http-tests/src/test/scala/akka/http/scaladsl/server/TestServer.scala

```diff
@@ -14,11 +14,90 @@ import akka.stream._
 import akka.stream.scaladsl._
 import akka.http.scaladsl.Http
 import akka.http.scaladsl.common.EntityStreamingSupport
+import akka.http.scaladsl.model.headers.{ HttpOrigin, `Access-Control-Allow-Methods`, `Access-Control-Allow-Origin` }
+import akka.http.scaladsl.server.util.ApplyConverter
 
 import scala.concurrent.ExecutionContext
 import scala.concurrent.duration._
 import scala.io.StdIn
 
+object SwaggerRoute {
+  import akka.http.scaladsl.marshallers.sprayjson.SprayJsonSupport._
+
+  import spray.json.DefaultJsonProtocol._
+
+  case class ParameterSpec(
+    name: String,
+    in:   String
+  )
+  case class ResponseSpec(
+    summary: String
+  )
+  case class OperationSpec(
+    summary:    String,
+    parameters: Seq[ParameterSpec]        = Vector.empty,
+    responses:  Map[String, ResponseSpec] = Map.empty
+  )
+  case class PathSpec(
+    get:  Option[OperationSpec] = None,
+    post: Option[OperationSpec] = None,
+    put:  Option[OperationSpec] = None
+  )
+
+  case class OpenApi(
+    title:   String,
+    openapi: String,
+    paths:   Map[String, PathSpec]
+  )
+  object OpenApi {
+    implicit val paramaterSpecFormat = jsonFormat2(ParameterSpec.apply _)
+    implicit val responseSpecFormat = jsonFormat1(ResponseSpec.apply _)
+    implicit val operationSpecFormat = jsonFormat3(OperationSpec.apply _)
+    implicit val pathSpecFormat = jsonFormat3(PathSpec.apply _)
+    implicit val openApiFormat = jsonFormat3(OpenApi.apply _)
+  }
+
+  import Directives._
+  def route(forRoute: Route): Route =
+    complete(apiDescForRoute(forRoute))
+
+  private def apiDescForRoute(route: Route): OpenApi = {
+    case class RoutePath(segments: Seq[DirectiveRoute], last: Route) {
```

Comment:
So, what we get here per leaf route is an chain ordered chain of all (the metadata of) the directives that are applied for that leaf route. Once we expose enough metadata, we should be able to generate openapi specifications but also do "static" (= at route building time) analytics of the routing tree. Even with this simple prototype we could prevent that people would be accidentally nesting `path` or method directives.